h2o-py/h2o/explanation/_explain.py

```diff
@@ -1349,14 +1360,25 @@ def _consolidate_varimps(model):
     consolidated_varimps = {k: v for k, v in varimp.items() if k in x}
     to_process = {k: v for k, v in varimp.items() if k not in x}
 
+    domain_mapping = _get_domain_mapping(model)
+    encoded_cols = ["{}.{}".format(name, domain)
+                    for name, domains in domain_mapping.items()
+                    if domains is not None
+                    for domain in domains + ["missing(NA)"]]
+    if len(encoded_cols) > len(set(encoded_cols)):
+        for x in set(encoded_cols):
+            encoded_cols.remove(x)
+        raise RuntimeError("Ambiguous encoding of the column x category pairs: {}".format(set(encoded_cols)))
```

Comment:
So we prefer raising an error rather than fixing the encoding syntax? Interesting…

A simple compatible way would be to externalize the encoding format, ie. expose it as a global variable:
```
varimp_col_format="{name}[{domain}]"
```
and tell users that they should set `h2o.explanation.varimp_col_format="{name}.{domain}"` if they have compatibility issues...
@michalkurka ?

The problem with raising an error here, is that is doesn't really help the user: are we basically asking him to rename the columns of his training data and retrain his model? This is asking a lot for something that can be fixed easily directly on client side.

Reply:
> This is asking a lot for something that can be fixed directly on client side.

@sebhrusen I don't think this could be solved on client side. The problem is created on the backend -  we don't validate if the encoded name is unique.

Imagine situation where we have a categorical column `A` with categories `a`,`b`, and `a.b` and another categorical column `A.a` with categories `a`, and `b`. The backend will produce varimp for some algos with names like `A.a.b` and `A.a.b` and on client I have to decode it. Unfortunately, varimp is sorted by variable importance and not the actual order of `names`, so we cannot easily decode it on backend either.

Using `"{name}[{domain}]"` would have the same issues if we can have square brackets in the `name` and `domain`, e.g., `name = "A[b]", domain = "c"` would end up encoded the same way as `name = "A", domain ="b][c"`. And for cases like that, there's the exception.

Or I am missing something that can be done on client side for that?

Reply:
I think a proper solution is on the backend: the backend should be able to decode even ambiguous encodings because it knows the logic of how they were created (the order of the columns is critical for decoding ambiguous)

throwing an error doesn't seem like the best solution to me

I think the best solution here would be if varimp method also returned the original feature name as a new column in the output

we can keep this PR but turn the error into a warning instead - to inform user we are not doing it right (for now)

Reply:
@tomasfryda I agree that there's no unique syntax that will work universally: this is the reason why I suggest to give the possibility for the user to chose a format that will work for his data.

My preference for brackets instead of dots as the default is just a question of probability: it's much more common to have a single dot in a column name than `][`…
But I wouldn't mind keeping "{name}.{domain}" as the default.
My point is simply that if we detect a conflict, we could print a warning (or maybe even an error) suggesting to change  the global varimp format to avoid the conflict.

> I think the best solution here would be if varimp method also returned the original feature name as a new column in the output

yes, having raw information in addition to the encoded one is always better.

Reply:
Looking at XGBoostNativeVarImp parsing[1], I am not sure how easy it will be to get the original feature name in that case and also it looks to me that XGBoost merges the conflicting encoded variables:

```r
library(h2o)
df <- data.frame(y=rep_len(1:5, 1000))
df$a <- as.factor(c("b.c", "d"))
df$a.b <- as.factor(letters[1:5])
df$y <- df$y * c(1, 10)

xgb <- h2o.xgboost(y="y", training_frame = as.h2o(df))
h2o.varimp(xgb)
```
=>
```
Variable Importances: 
  variable relative_importance scaled_importance percentage
1    a.b.c       386479.750000          1.000000   0.702795
2    a.b.a        99022.585938          0.256217   0.180068
3    a.b.b        58419.015625          0.151157   0.106232
4    a.b.d         5996.760742          0.015516   0.010905
5    a.b.e            0.000029          0.000000   0.000000

```

```r
library(h2o)
df <- data.frame(y=rep_len(1:5, 1000))
df$a <- as.factor(c("b.c", "d"))
df$ab <- as.factor(letters[1:5]) # NOTE: the only difference is in the name on this line
df$y <- df$y * c(1, 10)

xgb <- h2o.xgboost(y="y", training_frame = as.h2o(df))
h2o.varimp(xgb)
```
=>
```
Variable Importances: 
  variable relative_importance scaled_importance percentage
1    a.b.c       359073.187500          1.000000   0.652958
2     ab.a        99022.585938          0.275773   0.180068
3     ab.b        58419.015625          0.162694   0.106232
4     ab.c        27406.578125          0.076326   0.049838
5     ab.d         5996.760742          0.016701   0.010905
6     ab.e            0.000029          0.000000   0.000000
```

Personally, I'd rather change the errors to warnings and then gradually improve the code base than getting stuck at trying to make the @michalkurka's solution (for that I'd create a new JIRA). Is that ok with @sebhrusen and @michalkurka ? 

[1] https://github.com/h2oai/h2o-3/blob/master/h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/predict/XGBoostNativeVariableImportance.java#L70-L99

Reply:
absolutely - the good first step is to issue a warning (basically in this PR just error -> warning)

This will give us time to figure out the next step and the proper solution.